src/trees/NodeIndex.h

```diff
@@ -33,117 +33,78 @@
 
 #include <iostream>
 
-#include "MRCPP/mrcpp_declarations.h"
-
 namespace mrcpp {
 
 template <int D> class NodeIndex final {
 public:
-    NodeIndex(int n = 0, const int *l = nullptr);
-    NodeIndex(const NodeIndex<D> &idx);
-    NodeIndex(const NodeIndex<D> &pIdx, int cIdx);
+    // regular constructors
+    NodeIndex(int n = 0, const std::array<int, D> &l = {})
+            : N(static_cast<short int>(n))
+            , L(l) {}
+
+    // relative constructors
+    NodeIndex<D> parent() const {
+        std::array<int, D> l;
+        for (int d = 0; d < D; d++) l[d] = (this->L[d] < 0) ? (this->L[d] - 1) / 2 : this->L[d] / 2;
+        return NodeIndex<D>(this->N - 1, l);
+    }
+    NodeIndex<D> child(int cIdx) const {
+        std::array<int, D> l;
+        for (int d = 0; d < D; d++) l[d] = (2 * this->L[d]) + ((cIdx >> d) & 1);
+        return NodeIndex<D>(this->N + 1, l);
+    }
 
-    inline NodeIndex<D> &operator=(const NodeIndex<D> &idx);
-    inline bool operator==(const NodeIndex<D> &idx) const;
-    inline bool operator!=(const NodeIndex<D> &idx) const;
+    // comparisons
+    bool operator!=(const NodeIndex<D> &idx) const { return not(*this == idx); }
+    bool operator==(const NodeIndex<D> &idx) const {
+        bool out = (this->N == idx.N) ? true : false;
+        for (int d = 0; d < D; d++) out &= (this->L[d] == idx.L[d]);
+        return out;
+    }
 
-    void setScale(int n) { this->N = (short int)n; }
-    inline void setTranslation(const int *l);
+    // setters
+    void setScale(int n) { this->N = static_cast<short int>(n); }
+    void setTranslation(const std::array<int, D> &l) { this->L = l; }
 
+    // value getters
     int getScale() const { return this->N; }
     int getTranslation(int d) const { return this->L[d]; }
-    int *getTranslation() { return this->L; }
-    const int *getTranslation() const { return this->L; }
 
+    // reference getters
+    int &operator[](int d) { return this->L[d]; }
+    const int &operator[](int d) const { return this->L[d]; }
+    std::array<int, D> &getTranslation() { return this->L; }
+    const std::array<int, D> &getTranslation() const { return this->L; }
+
+    // ostream printer
     friend std::ostream &operator<<(std::ostream &o, const NodeIndex<D> &idx) { return idx.print(o); }
```

Comment:
You can unfriend this one too. Make `print` a `public` method and declare the overload outside the class:
```c++
template <int D>
std::ostream &operator<<(std::ostream &o, const NodeIndex<D> &idx) { return idx.print(o); }
```